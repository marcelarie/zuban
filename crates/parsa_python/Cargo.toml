[package]
name = "parsa_python"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
publish = false
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
test_utils.workspace = true
