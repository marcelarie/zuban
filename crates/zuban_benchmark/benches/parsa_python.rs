use criterion::{criterion_group, criterion_main, Criterion};

use parsa_python_cst::Tree;

fn parse_file(copies: usize) {
    let some_code = utils::dedent(
        r#"
        import os
        def x(a, b, c):
            try:
                return a
            except Exception:
                return [1, 2, 3, 4, 5, 6]
        class C:
            x: int
            y = 1 # type: int
            z: dict[int, list[str]] = {1: ["", r''], 2: [f""" {1} """]}
            def f(self, x: list[str], y: int): ...
    "#,
    );
    assert_eq!(some_code.len(), 262);
    Tree::parse(some_code.repeat(copies).into());
}

fn bench_parser(c: &mut Criterion) {
    c.bench_function("parse big file", |b| b.iter(|| parse_file(1000)));
    c.bench_function("parse small file", |b| b.iter(|| parse_file(1)));
}

// Register the benchmarks
criterion_group!(benches, bench_parser);
criterion_main!(benches);
