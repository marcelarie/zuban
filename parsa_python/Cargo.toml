[package]
name = "parsa_python"
version = "0.1.0"
edition.workspace = true
rust-version.workspace = true
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
