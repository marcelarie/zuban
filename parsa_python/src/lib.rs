use parsa::{create_parser,create_node,create_token};


create_token!(PythonToken, enum TokenType, String, Number, Endmarker);
create_node!(PythonNode, enum NodeType, file_input, foo);

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
        parse_python("bar");
        dbg!(TokenType::get_map());
        dbg!(NodeType::get_map());
        return
    }
}

struct PythonTokenizer {
}

impl PythonTokenizer {
    fn new_tok(&self) -> PythonToken {
        PythonToken {
            start: 0,
            length: 0,
            type_: TokenType::Endmarker,
        }
    }
}

impl parsa::Tokenizer<PythonToken> for PythonTokenizer {
    fn new(_code: &str) -> Self {
        Self {}
    }

    fn yield_next(&self) -> PythonToken {
        self.new_tok()
    }

}

create_parser!(parse_python, PythonTree, PythonNode,
               PythonToken, PythonTokenizer, TokenType, NodeType);
//create_parser!(parse_pythonx, PythonTreex, PythonNodex,
//               PythonToken, PythonTokenizer, TokenType, NodeType);
