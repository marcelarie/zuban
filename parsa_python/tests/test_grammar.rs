extern crate insta;

use parsa_python::*;

fn dedent(s: &'static str) -> String {
    let mut indent = usize::MAX;
    let lines: &Vec<_> = &s.split('\n').collect();
    for line in lines {
        if !line.trim().is_empty() {
            indent = std::cmp::min(indent, line.len() - line.trim_start().len());
        }
    }
    if indent == usize::MAX {
        return s.to_string();
    }
    let new_lines: Vec<_> = lines
        .iter()
        .map(|line| {
            if line.len() >= indent {
                &line[indent..]
            } else {
                line
            }
        })
        .collect();
    new_lines.join("\n")
}

fn tree_to_string(tree: PyTree) -> String {
    fn recurse(code: &mut String, node: &PyNode, depth: usize) {
        *code += &" ".repeat(depth);
        *code += &format!(
            "{}: {}-{}{}\n",
            node.type_str(),
            node.start(),
            node.end(),
            if node.is_leaf() {
                format!(" {:?}", node.as_code())
            } else {
                "".to_string()
            }
        );
        for c in node.iter_children() {
            assert_eq!(node.index, c.parent().unwrap().index);
            recurse(code, &c, depth + 1);
        }
    }

    let root = tree.root_node();
    assert!(root.parent().is_none());
    let mut code = String::new();
    recurse(&mut code, &root, 0);
    code
}

macro_rules! parametrize_snapshots {
    ($($name:ident : $input:expr;)*) => {$(
        #[test]
        fn $name() {
            let tree = parse($input.into());
            insta::assert_snapshot!(stringify!($name), tree_to_string(tree));
        }
    )*}
}

parametrize_snapshots!(
    empty: "".to_owned();
    very_simple: "1".to_owned();
    json: "{foo: 1}\n".to_owned();
    simple: dedent("
        a(bar:=1, foo=1);
        if a:
         1
        else: 2
        ");
    assignments_simple: dedent("
        a = 1
        a, b = 2
        a, b, = 2
        (a, b) = 2
        ((a, b)) = 2
        a = b = 3
        a = (b, c) = 3
        a = (b, c,) = 3
        a = b, c, = 3
        ");
    assignments_star: dedent("
        *foo = *bar = *baz
        i, *foo = ((j, [*bar, k])) = *baz
        *foo, *bar = yield baz
        ");
    assignments_aug: dedent("
        foo += 3
        foo.bar |= 3
        foo, bar += 3
        ");
    assignments_setitem: dedent("
        foo[0] += 3
        foo[0] = a = 3
        foo[a, b] = c
        foo[a::, b:f, c:=3] = c
        ");
    assignments_annotation: dedent("
        foo: bar
        foo: int = 3
        foo.bar: baz
        foo: bar = a = 4
        ");
    assignments_with_call: dedent("
        foo().bar = 1
        foo(1,).bar[1] = 1
        foo(a for a in b).bar[1] = 1
        ");
    for_stmt: dedent("
        for x in [1,2]:
            ...
        for *b in *c: ...
        for c.x, (*b, d.y[0]) in i, *j, k: ...
        ");
    comprehension_simple: dedent("
        [1 for a in b]
        [a for *a, c in b if c if d]
        foo(a for (*a, (*c, b),), in d)
        ");
    comprehension_fails: dedent("
        # fails
        [1 for a in *b]
        [1 for a in b if a else c]
        ");
    del_stmt: dedent("
        del foo
        del foo,
        del foo, bar.baz
        del foo[1]
        ");
    cls: dedent("
        class Foo(object):
            def __init__(self, /, f, *, g):
                pass

            @property
            def prop(self):
                pass
        ");
    import_names: dedent("
        import foo
        import foo.bar as baz
        import foo, bar, baz
        import foo.bar.baz
        ");
    import_name_fails: dedent("
        import foo,
        import foo, bar,
        ");
    import_from_normal: dedent("
        from foo import bar
        from foo import bar as baz
        from foo.bar.baz import *
        ");
    import_from_dotted: dedent("
        from . import (foo as baz, bar)
        from . import (foo, bar,)
        from ..foo import bar as baz
        from ... import *
        ");
    import_from_fails: dedent("
        from . import foo,
        from . import foo, bar,
        ");
    calls: dedent("
        foo(bar)
        foo(bar_foo=3)
        foo(bar_baz:=2)
        foo(**kwargs)
        ");
    nested_call: dedent("
        foo(1)()
        ");
    complex_calls: dedent("
        foo(bar, 3)
        foo(bar, *foo)
        foo(bar, *foo, name=3)
        foo(bar, *foo, name=3, **kwargs)
        foo(bar, *foo, name=3, **kwargs, b=3)
        foo(**kwargs, b=3, **foo)
        ");
    operators: dedent("
        f | d | c + 5 & 2 * 3 * 4 ** 5 ** 6
        ");
    simple_error_recovery: dedent("
        a + 3 +
        b = 3
        ");
    terminal_error_recovery: dedent("
        ?
        ");
    terminal_and_nonterminal_error_recovery: dedent("
        foo
        (1) + ? + c
        bar
        ");
    keyword_recovery: dedent("
        foo
        def
        bar
        else:
        baz
        ");
    match_simple: dedent("
        match foo:
            case ['a']: ...
            case _:
                pass
        ");
    soft_keyword_underscore: dedent("
        match _:
            case ['a']: _(_)
            case _:
                _
        match = 3
        ");
    match_underscore_with_as: dedent("
        match foo:
            case bar as _: pass
        ");
    match_class_pattern: dedent("
        match a:
            case A():
                pass
            case B(1, 2):
                pass
            case B(1, b=2):
                pass
            case B(a=1, b=2):
                pass
        ");
    dict_literal1: dedent("
        {1: 2}
        {**foo}
        {**foo, 1: 2}
        {**foo, **bar}
        ");
    function_simple: dedent("
        def foo(): ...
        def foo(a: 3): ...
        def foo(a, b,): ...
        def foo(a: f = 3, b=4): ...
        def foo(a=3, b=4): ...
        ");
    function_args: dedent("
        def foo(a, *, b=3): ...
        def foo(a=3, *, c): ...
        def foo(a, *, x, **kwargs): ...
        def foo(a, *args: int, **fooo: str,): ...
        def foo(a, /, b: i, c: i=3, **kwargs): ...
        def foo(*args): ...
        ");
    lambda_simple: dedent("
        lambda: ...
        lambda a: ...
        lambda a, b,: ...
        lambda a=3, b=4: ...
        ");
    lambda_slash: dedent("
        lambda a, /, b=3: ...
        lambda a=3, /,: ...
        lambda a=3, /: ...
        lambda a=3, /, *args, c=3: ...
        lambda a=3, /, c=3: ...
        lambda a, /, b, c=3, **kwargs: ...
        lambda a, a=3, /, c=3, *, d: ...
        ");
    lambda_args: dedent("
        lambda a, *, b=3: ...
        lambda a=3, *, c: ...
        lambda a, *, x, **kwargs: ...
        lambda a, *args, **fooo ,: ...
        lambda *args: ...
        ");
    lambda_kwargs: dedent("
        lambda **kwargs: ...
        lambda a=3, **kwargs: ...
        lambda a, **kwargs: ...
        lambda a, /, b, *, x, **kwargs: ...
        ");
    lambda_fails: dedent("
        lambda a, *, **kwargs: ...
        lambda a=3, b: ...
        lambda **a, b: ...
        lambda a=3, *,: ...
        lambda ,: ...
        ");
    lambda_fail_with_slash: dedent("
        lambda a=3, /, b: ...
        lambda a=3, /, b=3, c: ...
        lambda a, /, b=3, c: ...
        lambda z, a=3, d, /, b: ...
        ");
    with_items_simple: dedent("
        with foo as bar: ...
        with foo as bar, baz: ...
        ");
    with_items_parentheses: dedent("
        with (foo as bar): ...
        with (foo as bar, baz): ...
        with ((foo)): ...
        with (i for i in 42): ...
        with (yield y): ...
        ");
    string_literals_in_strings: dedent(r#"
        '"""'
        "'''"
        ""
        ''
        "#);
    bytes: dedent(r#"
        b'asdf\n\''
        br'\asdf'
        Br'asdf"""'
        bR''
        BR'\''
        "#);
    exception_group: dedent(r#"
        try:
            ...
        except* SpamError:
            ...
        except* FooError as e:
            ...
        except* (BarError, BazError) as e:
            ...
        "#);
    invalid_syntax: dedent(r#"
        def f():
            (x,) += 1
            z
        "#);
);
