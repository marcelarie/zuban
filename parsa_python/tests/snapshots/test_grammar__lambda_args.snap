---
source: parsa_python/tests/test_grammar.rs
expression: tree_to_string(tree)
---
file: 0-144
 stmt: 1-23
  simple_stmts: 1-23
   simple_stmt: 1-22
    expr_stmt: 1-22
     star_expressions: 1-22
      expression: 1-22
       lambda: 1-22
        Keyword: 1-7 "lambda"
        lambda_parameters: 8-17
         Name: 8-9 "a"
         Keyword: 9-10 ","
         Keyword: 11-12 "*"
         Keyword: 12-13 ","
         Name: 14-15 "b"
         Keyword: 15-16 "="
         expression: 16-17
          bitwise_or: 16-17
           Number: 16-17 "3"
        Keyword: 17-18 ":"
        expression: 19-22
         bitwise_or: 19-22
          Keyword: 19-22 "..."
   Newline: 22-23 "\n"
 stmt: 23-43
  simple_stmts: 23-43
   simple_stmt: 23-42
    expr_stmt: 23-42
     star_expressions: 23-42
      expression: 23-42
       lambda: 23-42
        Keyword: 23-29 "lambda"
        lambda_parameters: 30-37
         Name: 30-31 "a"
         Keyword: 31-32 "="
         expression: 32-33
          bitwise_or: 32-33
           Number: 32-33 "3"
         Keyword: 33-34 ","
         Keyword: 35-36 "*"
         Keyword: 36-37 ","
        Keyword: 37-38 ":"
        expression: 39-42
         bitwise_or: 39-42
          Keyword: 39-42 "..."
   Newline: 42-43 "\n"
 stmt: 43-65
  simple_stmts: 43-65
   simple_stmt: 43-64
    expr_stmt: 43-64
     star_expressions: 43-64
      expression: 43-64
       lambda: 43-64
        Keyword: 43-49 "lambda"
        lambda_parameters: 50-59
         Name: 50-51 "a"
         Keyword: 51-52 "="
         expression: 52-53
          bitwise_or: 52-53
           Number: 52-53 "3"
         Keyword: 53-54 ","
         Keyword: 55-56 "*"
         Keyword: 56-57 ","
         Name: 58-59 "c"
        Keyword: 59-60 ":"
        expression: 61-64
         bitwise_or: 61-64
          Keyword: 61-64 "..."
   Newline: 64-65 "\n"
 stmt: 65-95
  simple_stmts: 65-95
   simple_stmt: 65-94
    expr_stmt: 65-94
     star_expressions: 65-94
      expression: 65-94
       lambda: 65-94
        Keyword: 65-71 "lambda"
        lambda_parameters: 72-89
         Name: 72-73 "a"
         Keyword: 73-74 ","
         Keyword: 75-76 "*"
         Keyword: 76-77 ","
         Name: 78-79 "x"
         Keyword: 79-80 ","
         Keyword: 81-83 "**"
         Name: 83-89 "kwargs"
        Keyword: 89-90 ":"
        expression: 91-94
         bitwise_or: 91-94
          Keyword: 91-94 "..."
   Newline: 94-95 "\n"
 stmt: 95-126
  simple_stmts: 95-126
   simple_stmt: 95-125
    expr_stmt: 95-125
     star_expressions: 95-125
      expression: 95-125
       lambda: 95-125
        Keyword: 95-101 "lambda"
        lambda_parameters: 102-120
         Name: 102-103 "a"
         Keyword: 103-104 ","
         Keyword: 105-106 "*"
         Name: 106-110 "args"
         Keyword: 110-111 ","
         Keyword: 112-114 "**"
         Name: 114-120 "kwargs"
        Keyword: 120-121 ":"
        expression: 122-125
         bitwise_or: 122-125
          Keyword: 122-125 "..."
   Newline: 125-126 "\n"
 stmt: 126-144
  simple_stmts: 126-144
   simple_stmt: 126-143
    expr_stmt: 126-143
     star_expressions: 126-143
      expression: 126-143
       lambda: 126-143
        Keyword: 126-132 "lambda"
        lambda_parameters: 133-138
         Keyword: 133-134 "*"
         Name: 134-138 "args"
        Keyword: 138-139 ":"
        expression: 140-143
         bitwise_or: 140-143
          Keyword: 140-143 "..."
   Newline: 143-144 "\n"
 Endmarker: 144-144 ""

